### PR TITLE
feat(stdlib/universe): add interval package to window transformation

### PR DESCRIPTION
**Pull request by fchikwekwe**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: interval/window.go [hunk 1]
```diff
@@ -304,6 +312,3 @@ func isBeforeWithinMonth(a, b values.Time) bool {
 	}
 	return false
 }
-
-//TODO
-// Add tests very far away from the epoch
```
Comment (fchikwekwe) [inline interval/window.go:309]: Done in a previous PR.

File: stdlib/universe/range.go [hunk 2]
```diff
@@ -145,8 +143,15 @@ func createRangeTransformation(id execute.DatasetID, mode execute.AccumulationMo
 	d := execute.NewDataset(id, mode, cache)
 
 	bounds := a.StreamContext().Bounds()
+	e := execute.Bounds{}
+	if bounds != nil {
+		e = execute.Bounds{
+			Start: bounds.Start(),
+			Stop:  bounds.Stop(),
+		}
+	}
```
Comment (fchikwekwe) [inline stdlib/universe/range.go:152]: I added this check because I ran into panics when bounds was nil

File: stdlib/universe/window_test.go [hunk 3]
```diff
@@ -517,12 +514,12 @@ func TestFixedWindow_Process(t *testing.T) {
...
-							{start + 1*execute.Time(time.Minute), start + 2*execute.Time(time.Minute), start + execute.Time(80*time.Second), 8.0},
-							{start + 1*execute.Time(time.Minute), start + 2*execute.Time(time.Minute), start + execute.Time(90*time.Second), 9.0},
-							{start + 1*execute.Time(time.Minute), start + 2*execute.Time(time.Minute), start + execute.Time(100*time.Second), 10.0},
-							{start + 1*execute.Time(time.Minute), start + 2*execute.Time(time.Minute), start + execute.Time(110*time.Second), 11.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(0*time.Second), 0.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(10*time.Second), 1.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(20*time.Second), 2.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(30*time.Second), 3.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(40*time.Second), 4.0},
+							{start, start + 1*values.Time(time.Minute), start + values.Time(50*time.Second), 5.0},
```
Comment (fchikwekwe) [inline stdlib/universe/window_test.go:522]: Underlapping windows with a period that is smaller than the every now start at the beginning of the window. This was a breaking change, so tests dealing with this needed to be updated.

File: stdlib/universe/window_test.go [hunk 4]
```diff
@@ -568,12 +565,12 @@ func TestFixedWindow_Process(t *testing.T) {
...
```
Comment (fchikwekwe) [inline stdlib/universe/window_test.go:573]: Same thing here about underlapping windows.

File: execute/transformation.go [hunk 5]
```diff
@@ -87,7 +88,7 @@ func (ts TransformationSet) Finish(id DatasetID, err error) {
 // StreamContext represents necessary context for a single stream of
 // query data.
 type StreamContext interface {
-	Bounds() *Bounds
+	Bounds() *interval.Bounds
```
Comment (jsternberg) [inline execute/transformation.go:91]: This change isn't clear to me. It seems that `interval.Bounds` is intended to be created from the window function and that the `interval.Bounds` itself isn't meaningful outside of it being generated by the windowing function. I fear that using `interval.Bounds` everywhere where we need a start and stop and allowing a person to create a bounds will invalidate some of the math included in the bounds with the index and give a false sense of security. I wonder if we can keep this the same so `execute.Bounds` is used to denote a complete boundary (such as for range) while `interval.Bounds` is used to describe the boundary of an interval. Does that make sense?

File: interval/bounds.go [hunk 6]
```diff
@@ -22,6 +22,24 @@ type Bounds struct {
 	index int
 }
 
+// NewBounds create a new Bounds given start and stop values
+func NewBounds(start, stop values.Time) *Bounds {
```
Comment (jsternberg) [inline interval/bounds.go:26]: See comment above. I fear that this method may not work the way we want it to.

**Review by jsternberg — COMMENTED** [review]
The windowing portions look fine but I think the PR can be improved by removing the `interval.Bounds` related portions and using the original `execute.Bounds` in those places. I think the two structs potentially serve different purposes and both are useful. `execute.Bounds` describes a hard boundary for the query such as a start and stop while `interval.Bounds` represents a bounds produced by the windowing function. If we confuse those, I fear that we might run into improper use of the libraries since some of the interval logic relies on the index being present.

File: execute/executor.go [hunk 7]
```diff
@@ -44,13 +45,18 @@ func NewExecutor(logger *zap.Logger) Executor {
...
+	bounds        *interval.Bounds
+	executeBounds *Bounds
 }
 
-func (ctx streamContext) Bounds() *Bounds {
+func (ctx streamContext) Bounds() *interval.Bounds {
 	return ctx.bounds
 }
 
+func (ctx streamContext) ExecuteBounds() *Bounds {
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply execute/executor.go:56]: sure

File: interval/bounds.go [hunk 8]
```diff
@@ -22,6 +22,24 @@ type Bounds struct {
...
+	return &Bounds{
+		start: start,
+		stop:  stop,
+	}
+}
+
+// SetStart sets the start on a given Bounds
+func (b Bounds) SetStart(start values.Time) {
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply interval/bounds.go:34]: fair enough. I think I created them when I was still modifying too much code. They're not actually used in flux.

File: interval/bounds.go (hunk 6 above)
Comment (fchikwekwe) [reply interval/bounds.go:26]: I'm guessing you mean that creating a new bounds is akin to using this like in range? The actual use case is mainly for testing purposes. Do you think that's okay or do you still want me to remove it?

File: interval/bounds.go [hunk 9]
```diff
@@ -56,3 +79,53 @@ func (b Bounds) Length() values.Duration {
...
+	if o.stop < b.stop {
+		i.stop = o.stop
+	}
+
+	return i
+}
+
+// Union returns the smallest bounds which contain both input bounds.
+// It returns empty bounds if one of the input bounds are empty.
+func (b *Bounds) Union(o *Bounds) *Bounds {
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply interval/bounds.go:111]: got it

File: execute/executor.go [hunk 10]
```diff
@@ -44,13 +45,18 @@ func NewExecutor(logger *zap.Logger) Executor {
 }
 
 type streamContext struct {
-	bounds *Bounds
+	bounds        *interval.Bounds
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply execute/executor.go:48]: Constructing an interval.Bounds requires a method to create one. I have the NewBounds method that I added for this, but you were saying something about it not working properly. I think this is a case for keeping it if we can.

[the remark this answers is not in the document]
Comment (fchikwekwe) [reply execute/executor.go:48]: Also, if this is the case, then maybe we get rid of the other exposure method mentioned for execute Bounds?

File: execute/transformation.go (hunk 5 above)
Comment (fchikwekwe) [reply execute/transformation.go:91]: Yeah, I guess that makes some sense. A version of bounds that is mutable for creating range bounds and a version of bounds that are created for windowing and immutable. Is that right?

Comment (fchikwekwe) [reply execute/transformation.go:91]: I went ahead and took your suggesion here. Only windowing bounds should now be affected.

File: interpreter/interpreter_test.go [hunk 11]
```diff
@@ -858,8 +858,8 @@ func TestStack(t *testing.T) {
 			FunctionName: "window",
 			Location: ast.SourceLocation{
 				File:   "universe.flux",
-				Start:  ast.Position{Line: 135, Column: 12},
-				End:    ast.Position{Line: 135, Column: 49},
+				Start:  ast.Position{Line: 136, Column: 12},
+				End:    ast.Position{Line: 136, Column: 49},
```
Comment (fchikwekwe) [inline interpreter/interpreter_test.go:862]: I changed the `universe.flux` file, so the desired output changed.

File: stdlib/universe/window.go [hunk 12]
```diff
@@ -181,8 +181,11 @@ func createWindowTransformation(id execute.DatasetID, mode execute.AccumulationM
 	cache := execute.NewTableBuilderCache(a.Allocator())
 	d := execute.NewDataset(id, mode, cache)
 
-	bounds := a.StreamContext().Bounds()
-	if bounds == nil {
+	bounds := execute.Bounds{
+		Start: a.StreamContext().Bounds().Start,
+		Stop:  a.StreamContext().Bounds().Stop,
+	}
+	if bounds == (execute.Bounds{}) {
```
Comment (jsternberg) [inline stdlib/universe/window.go:188]: Does this part need to be changed? I'm a bit uncomfortable with comparing a bounds to the empty struct as a bounds of `[0, 0)` is technically different from nil.

File: stdlib/universe/universe.flux [hunk 13]
```diff
@@ -76,6 +76,7 @@ builtin tripleExponentialDerivative : (<-tables: [{B with _value: A}], n: int) =
 builtin union : (tables: [[A]]) => [A] where A: Record
 builtin unique : (<-tables: [A], ?column: string) => [A] where A: Record
 builtin window : (<-tables: [A], ?every: duration, ?period: duration, ?offset: duration, ?timeColumn: string, ?startColumn: string, ?stopColumn: string, ?createEmpty: bool) => [B] where A: Record, B: Record
+builtin windowInterval : (<-tables: [A], ?every: duration, ?period: duration, ?offset: duration, ?timeColumn: string, ?startColumn: string, ?stopColumn: string, ?createEmpty: bool) => [B] where A: Record, B: Record
```
Comment (jsternberg) [inline stdlib/universe/universe.flux:79]: I thought that we were introducing this through a planner rule. Is this just for testing or is this a new function?

Comment (fchikwekwe) [reply stdlib/universe/universe.flux:79]: Yes, I needed to have a separate procedure spec for the planner to be aware of the new transformation. It required so much duplication that I decided to temporarily add a new transformation (windowInterval). We can delete the old window transformation and promote this one once the feature flag for interval is rolled out.

Comment (fchikwekwe) [reply stdlib/universe/universe.flux:79]: my WIP for the new feature flag https://github.com/influxdata/idpe/pull/9730

File: stdlib/universe/window.go (hunk 12 above)
Comment (fchikwekwe) [reply stdlib/universe/window.go]: I changed it so that the check is performed before the assignment. Does that look better to you?

Comment (fchikwekwe) [reply stdlib/universe/window.go]: It compares to `nil` now.

File: stdlib/universe/universe.flux (hunk 13 above)
Comment (jsternberg) [reply stdlib/universe/universe.flux:79]: If the goal is to replace it with a new procedure spec, this isn't required. The function name only needs to be specified if you plan on invoking it from directly within a flux script like `|> windowInterval`. If the plan is just to replace the window transformation with a new procedure spec, then you can omit this.

**Review by jsternberg — APPROVED** [review]

File: stdlib/universe/universe.flux (hunk 13 above)
Comment (fchikwekwe) [reply stdlib/universe/universe.flux]: I've removed the unnecesary pieces from the windowInterval transformation so that its not user facing any longer.

**Pull request merged**